trie/codec.go

```diff
@@ -20,15 +20,58 @@ func KeyEncode(k []byte) []byte {
 
 // keyToHex turns bytes into nibbles
 func keyToHex(in []byte) []byte {
-	l := len(in)*2 + 1
+	l := len(in) * 2
 	res := make([]byte, l)
 	for i, b := range in {
 		res[2*i] = b / 16
 		res[2*i+1] = b % 16
 	}
 
-	// last index of nibble array is set to 16 due to the way branches are indexed
-	// branch at 0...15 points to possible children, branch at 16 is the value at the branch
-	res[l-1] = 16
 	return res
 }
+
+// hexToKey performs the opposite of keyToHex; turns nibbles back into bytes
+// removes last byte if length of input is odd (set to 16 if using keyToHex)
+func hexToKey(in []byte) []byte {
+	l := len(in) / 2
+	res := make([]byte, l)
+	for i := 0; i < len(in)-1; i = i + 2 {
+		res[i/2] = in[i+1] | in[i]<<4
+	}
+	return res
+}
+
+// bigKeySize returns the node type's BigKeySize
+// BigKeySize is 125 if node is extension, 126 if node is leaf
+func bigKeySize(n node) int {
+	switch n.(type) {
+	case *leaf:
+		return 126
+	default:
+		return -1
+	}
+}
+
+// getPrefix returns the node type's prefix, used for encoding the node
+func getPrefix(n node) (prefix byte) {
+	switch n := n.(type) {
+	case *leaf:
+		return 1
+	case *branch:
+		if n.value == nil {
+			// branch without value
+			return 254
+		}
+		// branch with value
+		return 255
+	default:
+		return 0
+	}
+}
+
+func uint16ToBytes(in uint16) (out []byte) {
```

Comment:
I think we could move this into a common/bitutils, a package for utility functions relevant to bytes. Given it's only used by the trie package and not a core function it could be used elsewhere in the application and we won't want to pull in the entire trie package to have to use it.

Reply:
sure yeah